### ISSUE TITLE
v3 plugin protocol bug

### ISSUE THREAD

**Issue by Somefive** [helm/helm#5934]
Output of `helm version`:
version.BuildInfo{Version:"v3.0+unreleased", GitCommit:"6b96d2158b0db94f1e1792df2a0e5e7dfa6674b0", GitTreeState:"dirty", GoVersion:"go1.12.5"}

Output of `kubectl version`:
This issue is not directly related to kubectl's version.

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
Minikube. But again, seems not relative.

The bug happens to the use of plugins. In latest helm version 3 branch (the dev-v3), in file `src/helm/pkg/repo/chartrepo.go`, around Line 124, currently in this branch, it writes
```go
g, err := getter.NewHTTPGetter(indexURL, r.Config.CertFile, r.Config.KeyFile, r.Config.CAFile)
if err != nil {
	return err
}
resp, err := g.Get(indexURL)
```
which I think should be 
```go
if g, isHTTPGetter := r.Client.(*getter.HTTPGetter); isHTTPGetter {
	g.SetBasicAuth(r.Config.Username, r.Config.Password)
}
resp, err := r.Client.Get(indexURL)
```
to enable protocols provided in plugins. Currently, the code here is not using `client` retrieved in `func NewChartRepository` which will ignore plugins.

**Comment by hickeyma** [comment]
@Somefive Could you provide the plugin commands that are failing and the errors returned? 
I have been unable to reproduce when doing the following testing:
```console
$ helm version
version.BuildInfo{Version:"v3.0+unreleased", GitCommit:"6b96d2158b0db94f1e1792df2a0e5e7dfa6674b0", GitTreeState:"clean", GoVersion:"go1.12.1"}

$ helm plugin list
NAME	VERSION	DESCRIPTION

$ helm plugin install https://github.com/IBM/helm-stop
Installed plugin: stop

$ helm plugin list
NAME	VERSION 	DESCRIPTION                           
stop	0.1.0   	This plugin stops a release pods

$ helm plugin install https://github.com/adamreese/helm-last
Installed plugin: last

$ helm plugin list
NAME	VERSION 	DESCRIPTION                           
last	0.1.0   	get the last release name             
stop	0.1.0   	This plugin stops a release pods
```

**Comment by bacongobbler** [comment]
I can take a look at this as I'm currently re-working the getter package. After pouring over the code for the last week it would require some internal refactoring in order for helm plugins to pull from a chart repository, so I'm marking this as a feature. @Somefive has pointed out one area, but there are a few others (like in `pkg/downloader`) where we'd need to address this as well.

https://github.com/helm/helm/pull/5922 partially addresses this.

**Comment by bacongobbler** [comment]
@hickeyma this is in reference to allowing plugins registered as protocol downloaders for fetching charts. Have a look here: https://helm.sh/docs/plugins/#downloader-plugins

**Comment by hickeyma** [comment]
Thanks for "heads up" @bacongobbler.

**Comment by Somefive** [comment]
@hickeyma The problem happens to plugins providing other protocols. For me here, I was using Alibaba's **acr** protocol and the error happens to `g.Get` function and returns `unsupported protocol scheme`.
As @bacongobbler said, I also find there is some other TODO which will cause similar problems.
The fix here I provided is effective to `helm repo add` command but for `helm pull` or other things, this fix is far from enough. But since the `helm push` plugin is partially independent with `helm v3`, it is not affected too much once I can add repo.

**Comment by bacongobbler** [comment]
@Somefive if you have some time, would you mind going through and finding any gaps we need to address to make your use case possible? I'm going to be away for 3 weeks so I won't be able to work on anything any time soon, but I'd love to see any proposals to support this use case!

**Comment by Somefive** [comment]
OK, look into it and see if there's anything I can help.

**Comment by Somefive** [comment]
@bacongobbler I think the use case described in this issue has been solved in the latest version, aka v3.0.0-beta.2.  Although *user-agent* remains to be fixed, my case does not need to care that so this issue is okay now. Thanks for your support!

**Comment by bacongobbler** [comment]
Just getting back from vacation. Thanks for checking back in and I'm glad we were able to resolve your use case! Let us know if something else pops up. Thanks!
